src/main/java/io/github/thebusybiscuit/cscorelib2/blocks/BlockPosition.java

```diff
@@ -0,0 +1,179 @@
+import org.bukkit.World;
+import org.bukkit.Chunk;
+import org.bukkit.block.Block;
+
+import java.lang.ref.WeakReference;
+import java.util.Objects;
+
+/**
+ * This is the position of a block in a World. Using this class as opposed to {@link Location} is much better because it
+ * has a lower memory footprint and doesn't contain useless data such as yaw and pitch.<br />
+ * This is 12 bytes in memory whereas {@link Location} is 36 bytes.
+ * (Note, this is not accounting for things like object headers)
+ */
+public final class BlockPosition {
+
+    private final WeakReference<World> world;
+    private final long position;
+
+    public BlockPosition(World world, long position) {
+        Objects.requireNonNull(world, "World cannot be null!");
+
+        this.world = new WeakReference<>(world);
+        this.position = position;
+    }
+
+    public BlockPosition(World world, int x, int y, int z) {
+        Objects.requireNonNull(world, "World cannot be null!");
+
+        this.world = new WeakReference<>(world);
+        this.position = ((long) (x & 0x3FFFFFF) << 38) | ((long) (z & 0x3FFFFFF) << 12) | (long) (y & 0xFFF);
+    }
+
+    /**
+     * Gets the {@link World} this block is in. If this {@link World} has been unloaded it will throw
+     * an {@link IllegalStateException}. This should be getting handled properly by yourself! <br />
+     * <b>Note: This is held as a weak reference!</b>
+     *
+     * @return The {@link World} for this block.
+     */
+    public World getWorld() {
+        final World world = this.world.get();
+        if (world == null) {
+            throw new IllegalStateException("The reference of this BlockPositions world has been cleared");
+        }
+
+        return world;
+    }
+
+    /**
+     * Gets the long position of this block. This is constructed of the x, y and z. <br />
+     * This is encoded as follows: {@code ((x & 0x3FFFFFF) << 38) | ((z & 0x3FFFFFF) << 12) | (y & 0xFFF)}
+     *
+     * @return The position of this block.
+     */
+    public long getPosition() {
+        return position;
+    }
+
+    /**
+     * Gets the x for this block.
+     *
+     * @return This blocks x coordinate.
+     */
+    public int getX() {
+        return (int) (this.position >> 38);
+    }
+
+    /**
+     * Gets the y for this block.
+     *
+     * @return This blocks y coordinate.
+     */
+    public int getY() {
+        return (int) (this.position & 0xFFF);
+    }
+
+    /**
+     * Gets the z for this block.
+     *
+     * @return This blocks z coordinate.
+     */
+    public int getZ() {
+        return (int) (this.position << 26 >> 38);
+    }
+
+    /**
+     * Gets the {@link Block} at this position. Note, Bukkit will create a new instance so if you can avoid doing this
+     * then do as it is a bit costly.
+     *
+     * @return The {@link Block} at this location.
+     */
+    public Block getBlock() {
+        return getChunk().getBlock((getX() & 0xF), getY(), (getZ() & 0xF));
```

Comment:
May I know why you're doing this rather than getLocation().getBlock() ?

Reply:
I wrote this ages ago but I believe it was for performance reasons